### DISCUSSION TITLE
IScheduleListener.JobDeleted problems with RAMJobStore and ADOJobStore with SQLite

### DISCUSSION THREAD

**Discussion by David-A-Blankenship** [quartznet/quartznet#1096] [Q&A]
I am setting up a scheduler where I want to know when my jobs are removed from the job store, so I can clean up some files on disk related to those jobs. I have implemented a IScheduleListener by inheriting from ShedulerListenerSupport and overriding JobDeleted. 

When run using the RAMJobStore, JobDeleted does not get called when all of the triggers associated with the job are finalized. TriggerFinalized is called, and I can potentially use this for what I need.

**Why is JobDeleted not getting called with the RAMJobStore?**

When run using the ADOJobStore with SQLite, JobDeleted does get called when all of the triggers associated with the job are finalized and the job is removed from the job store. The problem that I run into here is that I cannot query the scheduler during the event handler to get the job details. I get the following error:

Error while notifying SchedulerListener of job deletion.
Quartz.JobPersistenceException: Failure setting up connection.
 ---> code = Busy (5), message = System.Data.SQLite.SQLiteException (0x800007AF): database is locked

Here is the code for my JobDeleted event handler:

```
    Public Overrides Async Function JobDeleted(jobKey As JobKey, Optional cancellationToken As Threading.CancellationToken = Nothing) As Task
        Console.WriteLine(New String("*"c, 100))
        Console.WriteLine($"Job Deleted: {jobKey}")

        ' These apear to cause problems with the ADO job store.
        Dim jobDetail As IJobDetail = Await _Scheduler.GetJobDetail(jobKey)
        Dim jobData = jobDetail.JobDataMap
        Dim jobFile As String = jobData.GetString("File")
        Console.WriteLine($"Job Data: ")
        Console.WriteLine($"{vbTab}File: {jobFile}")

        'Return MyBase.JobDeleted(jobKey, cancellationToken)
    End Function
```

I get the same error while trying to query the scheduler within the TriggerFinalized handler.  

As suggested, I am using the default quartz.jobStore.lockHandler.type.

**Are there some configuration settings that allow me to query the scheduler from within an IScheduleListener event handler with SQLite**?
**Is this just a problem with SQLite?**

Any help would be appreciated.

**Comment by lahma** [comment]
Just reading the code and trying track what is happening in your case. 

> Why is JobDeleted not getting called with the RAMJobStore?

This might be a bug. Would require a full repro to investigate (or more time in hands).  If the case would be that you don't have any triggers associated the notification might be missed (that's my thinking without verification), logic is here:

https://github.com/quartznet/quartznet/blob/2fccbc1a17c3fc1bd6f7433f8aedc2c541a1fa46/src/Quartz/Simpl/RAMJobStore.cs#L311-L335

> Are there some configuration settings that allow me to query the scheduler from within an IScheduleListener event handler with SQLite?
Is this just a problem with SQLite?

If you are being notified here:

https://github.com/quartznet/quartznet/blob/2fccbc1a17c3fc1bd6f7433f8aedc2c541a1fa46/src/Quartz/Impl/AdoJobStore/JobStoreSupport.cs#L1317-L1320

That would mean that you are still in active DB transaction and thus the database is locked. (in SQLite's case). So this would require some sort of different logic than calling back the scheduler in the listener. I'd advice against doing such a dependency of scheduler -> listener -> scheduler unless really needed.

I'd suggest that you use the source version of Quartz or just do symbol debugging in your solution and check what's the behavior for job deletion.

**Comment by David-A-Blankenship** [comment]
Thank you for your response. The only difference in JobDeleted getting called is the job store that I have selected. All of the other code is the same, so I do get different behavior based on the job store. 

As for the DB transaction for SQLite, I noticed that being the issue, but the JobDeleted event only receives a JobKey. There is not a lot I can do with a JobKey. There is a lot I can do with IJobDetails.  What I find interesting about the code snippet that you included is that it looks like the JobKey is taken from the IJobDetails to be passed to the event handler.

 I did notice that the IJobContext includes a reference to the IScheduler, so I can query the scheduler when my job is executed to determine if it will ever fire again. While I would rather the job store supply me with this information since it is managing the lifetime of the job details in the job store, this should work.

```
    Public Async Function Execute(context As IJobExecutionContext) As Task Implements IJob.Execute
        Try
            Console.WriteLine(New String("*"c, 100))
            Console.WriteLine($"Job Executing: {context.JobDetail.Key}")
            Console.WriteLine($"{vbTab}Trigger Key: {context.Trigger.Key}")
            Console.WriteLine($"{vbTab}Start Time: {context.Trigger.StartTimeUtc}")
            Console.WriteLine($"{vbTab}End Time: {context.Trigger.EndTimeUtc}")
            Console.WriteLine($"{vbTab}Fire Time: {context.FireTimeUtc}")
            Console.WriteLine($"{vbTab}Current Schedule: {context.ScheduledFireTimeUtc}")
            Console.WriteLine($"{vbTab}Next Schedule: {context.Trigger.GetNextFireTimeUtc()}")
            Console.WriteLine($"{vbTab}Now: {DateTimeOffset.Now}")
            Dim jobData = context.JobDetail.JobDataMap
            Dim jobFile As String = jobData.GetString("File")
            Console.WriteLine($"{vbTab}File: {jobFile}")

            'Dim triggerData = context.Trigger.JobDataMap
            'Dim triggerFile As String = triggerData.GetString("File")
            'Console.WriteLine($"{vbTab}File: {triggerFile}")

            ' Check for context.Trigger.GetNextFireTimeUtc().HasValue() to see if this trigger is done.
            ' Check to see if there are more triggers for this job in the scheduler. If not then this job is done.
            If Not context.Trigger.GetNextFireTimeUtc().HasValue() Then
                Dim jobTriggers = Await context.Scheduler.GetTriggersOfJob(context.JobDetail.Key)
                Console.WriteLine($"Triggers of Job: {jobTriggers.Count}")
                For Each jobTrigger As ITrigger In jobTriggers
                    Console.WriteLine($"{vbTab}Trigger Key:{jobTrigger.Key}")
                Next
                If jobTriggers.Count = 1 Then
                    Console.WriteLine($"Job will be deleted")
                End If
            End If
        Catch ex As Exception
            Console.WriteLine(ex.Message)
            Dim e2 = New JobExecutionException(ex)
            e2.RefireImmediately = True
            Throw e2
        End Try
    End Function

End Class
```
